### ISSUE TITLE
-g0 option not handled optimally

### ISSUE THREAD

**Issue by pavsa** [ccache/ccache#368]
Hi,

Setup:
- using CCACHE_BASEDIR
- using CCACHE_DEPEND

I've noticed that having -g0 in gcc params causes pwd being hashed into common hash, despite -g0 meaning no debug symbols (the same as not including -g0). This causes such cached files to be unavailable for other users of the cache.

If -g is detected, it forces ```generating_debuginfo = true```, but there's no checking if it is -g0 or not, which turns off debugging info. Or am I mistaken?

https://github.com/ccache/ccache/blob/128109ebb7ffd2a2168677cae42d0e0200f0dbd7/src/ccache.c#L2623-L2640

Isn't this how it should be handled?
```
+               if (str_eq(argv[i], "-g0")) {
+                   continue;
+               }
                if (str_startswith(argv[i], "-g")) {
                        generating_debuginfo = true;

```

**Comment by afbjorklund** [comment]
I think you might be right, historically we have used either of `-O2 -g` or `-O0`...
i.e. without debugging symbols was the default, so there was no need to disable

Patch above sounds reasonable, combine with if below and make a PR ?
``` C
if (str_startswith(argv[i], "-g") && !str_eq(argv[i], "-g0")) {
```

**Comment by jrosdahl** [comment]
@pavsa: Thanks for noticing this. Would you be interested in fixing it? Otherwise I can have a look at it.

I suggest making sure that this case is handled as well: `gcc -g3 -g0 ...`. I.e., the last `-g` option "wins". For extra niceness, add a test case.

**Comment by pavsa** [comment]
I will look at it tomorrow and I'll add a test case. Thank you.

**Comment by hwti** [comment]
Shouldn't `-g1`/`-g2` disable `generating_debuginfo_level_3` ?

`-gdwarf-3` sets `generating_debuginfo_level_3`, but it has nothing to do with the level.

Also, I'm skeptical about `using_split_dwarf`, in https://github.com/ccache/ccache/issues/149 my test showed that at least for clang, `-gsplit-dwarf -g1` means `-g1` (so ccache would expect a dwo when clang wouldn't generate one).

**Comment by jrosdahl** [comment]
> Shouldn't -g1/-g2 disable generating_debuginfo_level_3 ?

I deliberately did not do that for reasons discussed in https://github.com/ccache/ccache/pull/371#discussion_r271915027 (the _"-ggdb3 -g2 is not equivalent to -g3 -ggdb2"_ part) – letting the code reason too much about `-g*` flags is doomed to fail due to complex semantics.

> -gdwarf-3 sets generating_debuginfo_level_3, but it has nothing to do with the level.

Aha, I didn't know about that option. So ccache will currently assume debug level 3 for `-gdwarf-3`, which leads to potential cache misses but not incorrect results. Would you like to write an issue about improving the logic?

> Also, I'm skeptical about using_split_dwarf, in #149 my test showed that at least for clang, -gsplit-dwarf -g1 means -g1 (so ccache would expect a dwo when clang wouldn't generate one).

OK, that's worse:

```
% rm -f empty*; touch empty.c; gcc -c -gsplit-dwarf -g1 empty.c; ls -1 empty*  
empty.c
empty.dwo
empty.o
% rm -f empty*; touch empty.c; clang -c -gsplit-dwarf -g1 empty.c; ls -1 empty*
empty.c
empty.o
```

So Clang is buggy in this case, at least if it wants to be a drop-in replacement for GCC. Would you like to create an issue for this?

**Comment by hwti** [comment]
> > Shouldn't -g1/-g2 disable generating_debuginfo_level_3 ?
> 
> I deliberately did not do that for reasons discussed in [#371 (comment)](https://github.com/ccache/ccache/pull/371#discussion_r271915027) (the _"-ggdb3 -g2 is not equivalent to -g3 -ggdb2"_ part) – letting the code reason too much about `-g*` flags is doomed to fail due to complex semantics.

For `-glevel` /  `-ggdblevel`, it seems to be the last option which wins (ie `-ggdb3 -g2` => `-g2`, `-g3 -ggdb2` => `-ggdb2` (But when there are other options it becomes more difficult).

For clang, `-g3` is the same as `-g2` (but since IIRC we always need run_second_cpp with clang, it isn't a real issue).


> > -gdwarf-3 sets generating_debuginfo_level_3, but it has nothing to do with the level.
> 
> Aha, I didn't know about that option. So ccache will currently assume debug level 3 for `-gdwarf-3`, which leads to potential cache misses but not incorrect results. Would you like to write an issue about improving the logic?

How could it lead to cache misses ?
I think the only effect is forcing run_second_cpp in a case where it isn't needed.
I created #394.

> ```
> % rm -f empty*; touch empty.c; gcc -c -gsplit-dwarf -g1 empty.c; ls -1 empty*  
> empty.c
> empty.dwo
> empty.o
> % rm -f empty*; touch empty.c; clang -c -gsplit-dwarf -g1 empty.c; ls -1 empty*
> empty.c
> empty.o
> ```
> 
> So Clang is buggy in this case, at least if it wants to be a drop-in replacement for GCC. Would you like to create an issue for this?

This was already in #149 and I'm not sure it could be fixed, but I created #393 so we can discuss it and at perhaps add something in the documentation.

**Comment by jrosdahl** [comment]
> How could it lead to cache misses ?
> I think the only effect is forcing run_second_cpp in a case where it isn't needed.

My bad, I was thinking of what `generating_debuginfo` implies, not `generating_debuginfo_level_3`.
